### ISSUE TITLE
CVE-2021-3121 (gogo/protobuf) remains in dependencies

### ISSUE THREAD

**Issue by mdbooth** [openshift/library-go#1055]
I've been trying to squash CVE-2021-3121 in a library which imports library-go. My test is:

```
$ git grep -l "if skippy < 0 {"
vendor/go.etcd.io/etcd/auth/authpb/auth.pb.go
vendor/go.etcd.io/etcd/etcdserver/etcdserverpb/etcdserver.pb.go
vendor/go.etcd.io/etcd/etcdserver/etcdserverpb/raft_internal.pb.go
vendor/go.etcd.io/etcd/etcdserver/etcdserverpb/rpc.pb.go
vendor/go.etcd.io/etcd/mvcc/mvccpb/kv.pb.go
vendor/go.etcd.io/etcd/raft/raftpb/raft.pb.go
vendor/k8s.io/apimachinery/pkg/api/resource/quantity_proto.go
```

The hit in apimachinery was missed when they bumped their gogo/protobuf dep, but there's now a PR to fix that: https://github.com/kubernetes/kubernetes/pull/101306

library-go currently pulls in

```
go.etcd.io/etcd v0.5.0-alpha.5.0.20200910180754-dd1b699fc489
```

which is included in etcd v3.4.14. etcd merged a gogo/protobuf fix in commit 1b300cd9e269c99b2045bff89e12318146a897d3 which is:

* On the v3.5 branch
* Not included in any tag

I'm not personally confident in trying to bump library-go's etcd dep between branches, especially to a non-stable branch, as I don't understand the implications of this.

The etcd commit which bumps gogo/protobuf makes it clear they don't believe there is an actual vulnerability there: https://github.com/etcd-io/etcd/commit/19f7c6ef3e674dc8de1a6e4b8c276e77c5be1cc5 . However, it would make verification across the product much simpler if we were able to eliminate all potentially vulnerable code.

I'm not giving this github issue security handling because everything in it is already public, and etcd specifically don't think it's an issue. It should be a useful tracker for operators trying to squash this CVE in their own codebases.

**Comment by joelsmith** [comment]
While updating etcd would bring in new versions of the files, the changes are cosmetic only. Two of the three changes to the protobuf compiler did not address actual security problems. To find vulnerable generated code, I use the following command:

```
find . -name '*.pb.go' | \
xargs  grep -l 'if skippy < 0' | \
xargs -r awk -e '/if skippy < 0/ {a=4} /if \(iNdEx \+ skippy\) > postIndex/ &&' \
        -e 'a>0 {print FILENAME " " FNR ": " $0 " // vulnerable to CVE-2021-3121"} {a--}'
```

Note that it may not find vulnerable hand-written code as in the case of `vendor/k8s.io/apimachinery/pkg/api/resource/quantity_proto.go`.

**Comment by mdbooth** [comment]
Thanks, Joel. I have found no technical description of this vulnerability anywhere. gogo/protobuf itself may as well be in hieroglyphics, and the commit message `skippy peanut butter` didn't give me any understanding of the actual issue. Consequently I've just been looking for any code which would have been changed by that commit.

I'm happy to take your word for it and steal your test, though! I guess we're just tracking the PR in apimachinery here, then.
